File: frontend/src/api/overview.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { fetchOverview } from './overview';

// Mock fetch globally with proper typing
const mockFetch = vi.fn();
globalThis.fetch = mockFetch;

describe('Overview API', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  describe('fetchOverview - happy path', () => {
    it('should fetch overview data from backend', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.5,
        avgMemoryPercent: 62.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(mockFetch).toHaveBeenCalledWith('/api/overview', undefined);
      expect(result).toEqual(mockOverview);
    });

    it('should return correctly structured overview data', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 5, total: 5 },
        unhealthyPods: 0,
        avgCpuPercent: 30.2,
        avgMemoryPercent: 55.8,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.nodes).toBeDefined();
      expect(result.nodes.ready).toBe(5);
      expect(result.nodes.total).toBe(5);
      expect(result.unhealthyPods).toBe(0);
      expect(result.avgCpuPercent).toBe(30.2);
      expect(result.avgMemoryPercent).toBe(55.8);
    });

    it('should handle zero values correctly', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 0, total: 0 },
        unhealthyPods: 0,
        avgCpuPercent: 0,
        avgMemoryPercent: 0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.nodes.ready).toBe(0);
      expect(result.nodes.total).toBe(0);
      expect(result.unhealthyPods).toBe(0);
      expect(result.avgCpuPercent).toBe(0);
      expect(result.avgMemoryPercent).toBe(0);
    });

    it('should handle high usage percentages', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 10, total: 10 },
        unhealthyPods: 5,
        avgCpuPercent: 95.7,
        avgMemoryPercent: 88.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.avgCpuPercent).toBeGreaterThanOrEqual(0);
      expect(result.avgCpuPercent).toBeLessThanOrEqual(100);
      expect(result.avgMemoryPercent).toBeGreaterThanOrEqual(0);
      expect(result.avgMemoryPercent).toBeLessThanOrEqual(100);
    });
  });

  describe('fetchOverview - with namespace filter', () => {
    it('should fetch overview data for specific namespace', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.5,
        avgMemoryPercent: 62.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview('default');

      // Assert
      expect(mockFetch).toHaveBeenCalledWith('/api/overview?namespace=default', undefined);
      expect(result).toEqual(mockOverview);
    });

    it('should fetch overview data for kube-system namespace', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 3, total: 3 },
        unhealthyPods: 0,
        avgCpuPercent: 20.0,
        avgMemoryPercent: 40.0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview('kube-system');

      // Assert
      expect(mockFetch).toHaveBeenCalledWith('/api/overview?namespace=kube-system', undefined);
      expect(result).toEqual(mockOverview);
    });

    it('should omit namespace parameter when undefined', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.5,
        avgMemoryPercent: 62.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(mockFetch).toHaveBeenCalledWith('/api/overview', undefined);
      expect(result).toEqual(mockOverview);
    });

    it('should handle empty string namespace', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.5,
        avgMemoryPercent: 62.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview('');

      // Assert
      // Empty namespace should fetch all namespaces (no query param)
      expect(mockFetch).toHaveBeenCalledWith('/api/overview', undefined);
      expect(result).toEqual(mockOverview);
    });
  });

  describe('fetchOverview - error cases', () => {
    it('should handle network errors gracefully', async () => {
      // Arrange
      mockFetch.mockRejectedValueOnce(new Error('Network error'));

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow('Network error');
    });

    it('should handle non-200 responses', async () => {
      // Arrange
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 500,
        statusText: 'Internal Server Error',
      });

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow();
    });

    it('should handle 404 not found', async () => {
      // Arrange
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 404,
        statusText: 'Not Found',
      });

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow();
    });

    it('should handle 403 forbidden', async () => {
      // Arrange
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 403,
        statusText: 'Forbidden',
      });

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow();
    });

    it('should handle invalid JSON response', async () => {
      // Arrange
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => {
          throw new Error('Invalid JSON');
        },
      });

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow('Invalid JSON');
    });

    it('should handle timeout scenarios', async () => {
      // Arrange
      mockFetch.mockImplementation(() => {
        return new Promise((_, reject) => {
          setTimeout(() => reject(new Error('Request timeout')), 100);
        });
      });

      // Act & Assert
      await expect(fetchOverview()).rejects.toThrow('Request timeout');
    });
  });

  describe('fetchOverview - edge cases', () => {
    it('should handle decimal percentages correctly', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.567,
        avgMemoryPercent: 62.345,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.avgCpuPercent).toBeCloseTo(45.567, 2);
      expect(result.avgMemoryPercent).toBeCloseTo(62.345, 2);
    });

    it('should handle large unhealthy pod counts', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 100, total: 100 },
        unhealthyPods: 999,
        avgCpuPercent: 50.0,
        avgMemoryPercent: 50.0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.unhealthyPods).toBe(999);
    });

    it('should handle ready nodes equal to total nodes', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 10, total: 10 },
        unhealthyPods: 0,
        avgCpuPercent: 30.0,
        avgMemoryPercent: 40.0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.nodes.ready).toBe(result.nodes.total);
    });

    it('should handle ready nodes less than total nodes', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 5, total: 10 },
        unhealthyPods: 3,
        avgCpuPercent: 60.0,
        avgMemoryPercent: 70.0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert
      expect(result.nodes.ready).toBeLessThan(result.nodes.total);
    });
  });

  describe('fetchOverview - type safety', () => {
    it('should return type-safe overview data', async () => {
      // Arrange
      const mockOverview = {
        nodes: { ready: 2, total: 3 },
        unhealthyPods: 1,
        avgCpuPercent: 45.5,
        avgMemoryPercent: 62.3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => mockOverview,
      });

      // Act
      const result = await fetchOverview();

      // Assert - TypeScript compile-time check
      const ready: number = result.nodes.ready;
      const total: number = result.nodes.total;
      const unhealthy: number = result.unhealthyPods;
      const cpu: number = result.avgCpuPercent;
      const memory: number = result.avgMemoryPercent;

      expect(ready).toBe(2);
      expect(total).toBe(3);
      expect(unhealthy).toBe(1);
      expect(cpu).toBe(45.5);
      expect(memory).toBe(62.3);
    });
  });
});

File: handlers/overview_test.go
package handlers

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// skipIfNoCluster skips the test if no Kubernetes cluster is available.
func skipIfNoCluster(t *testing.T) {
	t.Helper()
	if _, err := getKubernetesClient(); err != nil {
		t.Skipf("skipping: no k8s cluster available (%v)", err)
	}
}

// TestOverviewHandler tests the /api/overview endpoint
func TestOverviewHandler(t *testing.T) {
	t.Run("should return 200 OK with overview data", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		// In CI environment without cluster, 500 is acceptable
		// In cluster environment, 200 is expected
		if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusInternalServerError {
			t.Errorf("expected status 200 or 500, got %d", res.StatusCode)
		}

		// If 200, verify JSON response structure
		if res.StatusCode == http.StatusOK {
			var overview map[string]interface{}
			if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
				t.Fatalf("failed to decode response: %v", err)
			}
			if overview == nil {
				t.Error("expected overview object, got nil")
			}
		}
	})

	t.Run("should set correct content-type header", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		contentType := w.Header().Get("Content-Type")
		if contentType != "application/json" {
			t.Errorf("expected Content-Type 'application/json', got '%s'", contentType)
		}
	})

	t.Run("should reject non-GET methods", func(t *testing.T) {
		// Arrange
		methods := []string{http.MethodPost, http.MethodPut, http.MethodDelete, http.MethodPatch}

		for _, method := range methods {
			t.Run(method, func(t *testing.T) {
				req := httptest.NewRequest(method, "/api/overview", nil)
				w := httptest.NewRecorder()

				// Act
				OverviewHandler(w, req)

				// Assert
				res := w.Result()
				defer res.Body.Close()

				if res.StatusCode != http.StatusMethodNotAllowed {
					t.Errorf("expected status 405 for %s, got %d", method, res.StatusCode)
				}
			})
		}
	})

	t.Run("should return valid overview response structure", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		// Verify expected fields exist
		expectedFields := []string{"nodes", "unhealthyPods", "avgCpuPercent", "avgMemoryPercent"}
		for _, field := range expectedFields {
			if _, exists := overview[field]; !exists {
				t.Errorf("expected field '%s' in response, but not found", field)
			}
		}
	})

	t.Run("should return nodes with ready and total counts", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		// Verify nodes structure
		nodes, ok := overview["nodes"].(map[string]interface{})
		if !ok {
			t.Fatal("expected 'nodes' to be an object")
		}

		if _, exists := nodes["ready"]; !exists {
			t.Error("expected 'ready' field in nodes object")
		}

		if _, exists := nodes["total"]; !exists {
			t.Error("expected 'total' field in nodes object")
		}
	})

	t.Run("should return non-negative unhealthyPods count", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		unhealthyPods, ok := overview["unhealthyPods"].(float64)
		if !ok {
			t.Fatal("expected 'unhealthyPods' to be a number")
		}

		if unhealthyPods < 0 {
			t.Errorf("expected unhealthyPods >= 0, got %f", unhealthyPods)
		}
	})

	t.Run("should return avgCpuPercent between 0 and 100", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		avgCpu, ok := overview["avgCpuPercent"].(float64)
		if !ok {
			t.Fatal("expected 'avgCpuPercent' to be a number")
		}

		if avgCpu < 0 || avgCpu > 100 {
			t.Errorf("expected avgCpuPercent between 0-100, got %f", avgCpu)
		}
	})

	t.Run("should return avgMemoryPercent between 0 and 100", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		avgMemory, ok := overview["avgMemoryPercent"].(float64)
		if !ok {
			t.Fatal("expected 'avgMemoryPercent' to be a number")
		}

		if avgMemory < 0 || avgMemory > 100 {
			t.Errorf("expected avgMemoryPercent between 0-100, got %f", avgMemory)
		}
	})

	t.Run("should handle Kubernetes client errors gracefully", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		// Should either succeed or return error status
		// In TDD Red phase, this might fail or return 500 if client is not configured
		if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusInternalServerError {
			t.Errorf("expected status 200 or 500, got %d", res.StatusCode)
		}
	})
}

// TestOverviewHandlerWithNamespaceFilter tests namespace filtering
func TestOverviewHandlerWithNamespaceFilter(t *testing.T) {
	t.Run("should accept namespace query parameter", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview?namespace=default", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		// Should not fail with namespace parameter
		if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusInternalServerError {
			t.Errorf("expected status 200 or 500, got %d", res.StatusCode)
		}
	})

	t.Run("should filter pods by namespace when specified", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview?namespace=kube-system", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		// Response should still contain all required fields
		if _, exists := overview["unhealthyPods"]; !exists {
			t.Error("expected 'unhealthyPods' field even with namespace filter")
		}
	})

	t.Run("should return all namespaces data when namespace not specified", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		// Should contain data from all namespaces
		if overview["unhealthyPods"] == nil {
			t.Error("expected unhealthyPods data for all namespaces")
		}
	})

	t.Run("should handle empty namespace parameter", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview?namespace=", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		// Empty namespace should be treated as all namespaces
		if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusInternalServerError {
			t.Errorf("expected status 200 or 500, got %d", res.StatusCode)
		}
	})

	t.Run("should handle non-existent namespace gracefully", func(t *testing.T) {
		skipIfNoCluster(t)

		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview?namespace=non-existent-namespace", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		// Should return valid response even for non-existent namespace
		// Likely with zero values
		if overview["unhealthyPods"] == nil {
			t.Error("expected unhealthyPods field even for non-existent namespace")
		}
	})
}

// TestOverviewHandlerResponseFormat tests exact response format
func TestOverviewHandlerResponseFormat(t *testing.T) {
	t.Run("should match expected JSON structure", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview struct {
			Nodes struct {
				Ready int `json:"ready"`
				Total int `json:"total"`
			} `json:"nodes"`
			UnhealthyPods    int     `json:"unhealthyPods"`
			AvgCpuPercent    float64 `json:"avgCpuPercent"`
			AvgMemoryPercent float64 `json:"avgMemoryPercent"`
		}

		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response into expected structure: %v", err)
		}

		// Verify types are correct (compile-time check ensures this)
		t.Logf("Nodes: ready=%d, total=%d", overview.Nodes.Ready, overview.Nodes.Total)
		t.Logf("UnhealthyPods: %d", overview.UnhealthyPods)
		t.Logf("AvgCpuPercent: %.2f", overview.AvgCpuPercent)
		t.Logf("AvgMemoryPercent: %.2f", overview.AvgMemoryPercent)
	})

	t.Run("should return ready nodes less than or equal to total nodes", func(t *testing.T) {
		// Arrange
		req := httptest.NewRequest(http.MethodGet, "/api/overview", nil)
		w := httptest.NewRecorder()

		// Act
		OverviewHandler(w, req)

		// Assert
		res := w.Result()
		defer res.Body.Close()

		var overview struct {
			Nodes struct {
				Ready int `json:"ready"`
				Total int `json:"total"`
			} `json:"nodes"`
		}

		if err := json.NewDecoder(res.Body).Decode(&overview); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		if overview.Nodes.Ready > overview.Nodes.Total {
			t.Errorf("ready nodes (%d) cannot exceed total nodes (%d)",
				overview.Nodes.Ready, overview.Nodes.Total)
		}
	})
}

// TestCalculateResourceUsageFallback tests metric calculation with and without metrics-server data
func TestCalculateResourceUsageFallback(t *testing.T) {
	// Helper to create a test node with given capacity and allocatable
	makeNode := func(name string, cpuCapacity, cpuAllocatable, memCapacity, memAllocatable string) corev1.Node {
		return corev1.Node{
			ObjectMeta: metav1.ObjectMeta{Name: name},
			Status: corev1.NodeStatus{
				Capacity: corev1.ResourceList{
					corev1.ResourceCPU:    resource.MustParse(cpuCapacity),
					corev1.ResourceMemory: resource.MustParse(memCapacity),
				},
				Allocatable: corev1.ResourceList{
					corev1.ResourceCPU:    resource.MustParse(cpuAllocatable),
					corev1.ResourceMemory: resource.MustParse(memAllocatable),
				},
			},
		}
	}

	t.Run("should return 0 for empty nodes list", func(t *testing.T) {
		cpu, mem := calculateResourceUsage(nil, nil)
		if cpu != 0 || mem != 0 {
			t.Errorf("expected 0/0 for empty nodes, got %f/%f", cpu, mem)
		}
	})

	t.Run("should fall back to capacity-allocatable when metricsMap is nil", func(t *testing.T) {
		nodes := []corev1.Node{
			makeNode("node-1", "4000m", "3800m", "8Gi", "7Gi"),
		}
		cpu, mem := calculateResourceUsage(nodes, nil)
		// capacity - allocatable: 200m / 4000m = 5%, 1Gi / 8Gi = 12.5%
		if cpu < 4.9 || cpu > 5.1 {
			t.Errorf("expected ~5%% CPU fallback, got %f", cpu)
		}
		if mem < 12.0 || mem > 13.0 {
			t.Errorf("expected ~12.5%% memory fallback, got %f", mem)
		}
	})

	t.Run("should use real metrics when metricsMap is provided", func(t *testing.T) {
		nodes := []corev1.Node{
			makeNode("node-1", "4000m", "4000m", "8Gi", "8Gi"),
		}
		metricsMap := map[string]nodeMetricsUsage{
			"node-1": {
				cpuMillis:   2000, // 2000m / 4000m = 50%
				memoryBytes: 4 * 1024 * 1024 * 1024, // 4Gi / 8Gi = 50%
			},
		}
		cpu, mem := calculateResourceUsage(nodes, metricsMap)
		if cpu < 49.9 || cpu > 50.1 {
			t.Errorf("expected ~50%% CPU from metrics, got %f", cpu)
		}
		if mem < 49.9 || mem > 50.1 {
			t.Errorf("expected ~50%% memory from metrics, got %f", mem)
		}
	})

	t.Run("should fall back for nodes not in metricsMap", func(t *testing.T) {
		nodes := []corev1.Node{
			makeNode("node-1", "4000m", "4000m", "8Gi", "8Gi"),
			makeNode("node-2", "4000m", "3800m", "8Gi", "7Gi"),
		}
		// Only node-1 has metrics; node-2 should fall back
		metricsMap := map[string]nodeMetricsUsage{
			"node-1": {
				cpuMillis:   2000,
				memoryBytes: 4 * 1024 * 1024 * 1024,
			},
		}
		cpu, mem := calculateResourceUsage(nodes, metricsMap)
		// node-1: 2000m/4000m, node-2 fallback: 200m/4000m
		// total: 2200m/8000m = 27.5%
		if cpu < 27.0 || cpu > 28.0 {
			t.Errorf("expected ~27.5%% CPU mixed mode, got %f", cpu)
		}
		if mem < 1 {
			t.Errorf("expected positive memory mixed mode, got %f", mem)
		}
	})

	t.Run("should clamp values between 0 and 100", func(t *testing.T) {
		nodes := []corev1.Node{
			makeNode("node-1", "1000m", "1000m", "1Gi", "1Gi"),
		}
		metricsMap := map[string]nodeMetricsUsage{
			"node-1": {
				cpuMillis:   5000, // 500% — should be clamped to 100
				memoryBytes: 10 * 1024 * 1024 * 1024,
			},
		}
		cpu, mem := calculateResourceUsage(nodes, metricsMap)
		if cpu != 100 {
			t.Errorf("expected CPU clamped to 100, got %f", cpu)
		}
		if mem != 100 {
			t.Errorf("expected memory clamped to 100, got %f", mem)
		}
	})
}

// TestCalculateNodeResourceUsageFallback tests per-node metric calculation
func TestCalculateNodeResourceUsageFallback(t *testing.T) {
	node := corev1.Node{
		ObjectMeta: metav1.ObjectMeta{Name: "test-node"},
		Status: corev1.NodeStatus{
			Capacity: corev1.ResourceList{
				corev1.ResourceCPU:    resource.MustParse("4000m"),
				corev1.ResourceMemory: resource.MustParse("8Gi"),
			},
			Allocatable: corev1.ResourceList{
				corev1.ResourceCPU:    resource.MustParse("4000m"),
				corev1.ResourceMemory: resource.MustParse("8Gi"),
			},
		},
	}

	t.Run("should return 0 when no metrics and capacity equals allocatable", func(t *testing.T) {
		cpu, mem := calculateNodeResourceUsage(node, nil)
		if cpu != 0 || mem != 0 {
			t.Errorf("expected 0/0 with nil metrics and equal capacity/allocatable, got %f/%f", cpu, mem)
		}
	})

	t.Run("should use real metrics when available", func(t *testing.T) {
		metricsMap := map[string]nodeMetricsUsage{
			"test-node": {
				cpuMillis:   1000,
				memoryBytes: 2 * 1024 * 1024 * 1024,
			},
		}
		cpu, mem := calculateNodeResourceUsage(node, metricsMap)
		if cpu < 24.9 || cpu > 25.1 {
			t.Errorf("expected ~25%% CPU, got %f", cpu)
		}
		if mem < 24.9 || mem > 25.1 {
			t.Errorf("expected ~25%% memory, got %f", mem)
		}
	})
}
